api/devicerepository.proto

```diff
@@ -303,11 +307,51 @@ message ListEndDeviceModelsResponse {
   repeated EndDeviceModel models = 1;
 }
 
-message MessagePayloadFormatter {
+message EncodedMessagePayload {
+  uint32 f_port = 1 [(validate.rules).uint32.lte = 255];
+  bytes frm_payload = 2 [(gogoproto.customname) = "FRMPayload"];
+  repeated string warnings = 3;
+  repeated string errors = 4;
+}
+
+message DecodedMessagePayload {
+  google.protobuf.Struct data = 1;
+  repeated string warnings = 2;
+  repeated string errors = 3;
+}
+
+message MessagePayloadDecoder {
   // Payload formatter type.
   PayloadFormatter formatter = 1 [(validate.rules).enum.defined_only = true];
   // Parameter for the formatter, must be set together.
   string formatter_parameter = 2;
```

Comment:
While we're at it, should we set a maximum length here?

Reply:
These are sourced from the device repository https://github.com/TheThingsNetwork/lorawan-devices/blob/master/lib/draft/schema.json#L240-L250 where no maximum length is enforced.

There is already an issue to add validation rules for the device repository API (https://github.com/TheThingsNetwork/lorawan-stack/issues/3694) for other message fields as well.

I am adding validation rules while I'm at it, but the proper place for these rules would be the schema of the Device Repository.

Reply:
Only the code files themselves don't have a length check.

FPort and FRMPayload length are validated by the device repository.

There is (currently) no limitation of the number of examples, the length of the example description and the elements in the input and output fields.

I tend to agree that we should keep the validation in the device repository. This API is only returning fields anyway.